lib/impl/refactorRules.js

```diff
@@ -0,0 +1,384 @@
+            var targetSectionId = targetSection && targetSection._id.toString();
+            return targetSectionId !== sectionIds[0];
+          });
+          if (targetInDifferrentSection) {
+            rulesFlaggedForDeletion[fieldRule._id.toString()] = true;
+          }
+        }
+      });
+    }
+  }
+
+  function updateOrDeleteRules(cb) {
+    //Updating each of the rules to account for any fields removed. Can't just simply delete them.
+
+    async.parallel([
+      function updateFieldRulesInDb(cb) {
+        async.each(form.fieldRules, function(formFieldRule, cb) {
+          var fieldRuleId = formFieldRule._id.toString();
+          fieldRuleModel.findOne({"_id": fieldRuleId}, function(err, fieldRule) {
+            if (err) {
+              return cb(err);
+            }
+
+            if (fieldRule === null) {
+              return cb(new Error("No field rule with id " + fieldRuleId + " found."));
+            }
+
+            //Update the ruleConditionalStatements and targets of the rule.
+            fieldRule.ruleConditionalStatements = formFieldRule.ruleConditionalStatements;
+            fieldRule.targetField = formFieldRule.targetField;
+            fieldRule.save(cb);
+          });
+        }, cb);
+      },
+      function updatePageRulesInDb(cb) {
+        async.each(form.pageRules, function(formPageRule, cb) {
+          var pageRuleId = formPageRule._id.toString();
+          pageRuleModel.findOne({"_id": pageRuleId}, function(err, pageRule) {
+            if (err) {
+              return cb(err);
+            }
+
+            if (pageRule === null) {
+              return cb(new Error("No page rule with id " + pageRuleId + " found."));
+            }
+
+            //Update the ruleConditionalStatements and targets of the rule.
+            pageRule.ruleConditionalStatements = formPageRule.ruleConditionalStatements;
+            pageRule.save(cb);
+          });
+        }, cb);
+      },
+      function deleteFieldRules(cb) {
+        async.eachSeries(fieldRulesToDelete, function(fieldRuleId, cb) {
+          fieldRuleModel.findOne({"_id": fieldRuleId}, function(err, fieldRule) {
+            if (err) {
+              return cb(err);
+            }
+
+            if (fieldRule !== null) {
+              fieldRule.remove(cb);
+            } else {
+              return cb();
+            }
+          });
+        }, cb);
+      },
+      function deletePageRules(cb) {
+        async.eachSeries(pageRulesToDelete, function(pageRuleId, cb) {
+          pageRuleModel.findOne({"_id": pageRuleId}, function(err, pageRule) {
+            if (err) {
+              return cb(err);
+            }
+
+            if (pageRule !== null) {
+              pageRule.remove(cb);
+            } else {
+              return cb();
+            }
+          });
+        }, cb);
+      }
+    ], cb);
+  }
+
+  //Need to delete any rules that are to be deleted
+  function updateFormRules(cb) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Function has a complexity of 23 which is greater than 20 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)